text/0026-serialization.md

```diff
@@ -0,0 +1,242 @@
+
+**In editor mode and development builds**, however, calling these functions records a watermark point, and any attempt to read or write past the watermark point will throw an exception. This ensures these functions are used properly, while avoiding the performance cost of per-operation checking in production builds.
+
+## Bitwise Reading and Writing
+
+Writing values in sizes measured in bits rather than bytes comes with a cost - first, it comes with a cost of having to track bitwise lengths and convert them to bytewise lenghts, and second, it comes with a cost of having to remember to add padding after your bitwise writes and reads to ensure the next bytewise write or read functions correctly, and to make sure the buffer length includes any trailing bits.
+
+To address that, `FastBufferReader` and `FastBufferWriter` do not, themselves, have bitwise operations. When needed, however, you can create a `BitwiseWriter` or `BitwiseReader` instance, which is an IDisposable to ensure that no unaligned bits are left at the end - from the perspective of `FastBufferReader` and `FastBufferWriter`, only bytes are allowed. `BitwiseWriter` and `BitwiseReader` operate directly on the underlying buffer, so calling non-bitwise operations within a bitwise context is an error (and will raise an exception in non-production builds).
+
+```
+FastBufferWriter writer = new FastBufferWriter(256, Allocator.TempJob);
+using(var bitWriter = writer.EnterBitwiseContext())
+{
+	bitWriter.WriteBit(a);
+	bitWriter.WriteBits(b, 5);
+} // Dispose automatically adds 2 more 0 bits to pad to the next byte.
+```
+
+## INetworkSerializable
+
+A new `BufferSerializer` class wrapping `FastBufferReader` and `FastBufferWriter` will replace the existing `NetworkSerializer` to provide the same two-way interface as the existing `NetworkSerializer`, and is rewritten as a **ref struct** so that no allocation overhead is incurred in calling the `INetworkSerializable`'s `Serialize()` method. A **ref struct** is chosen instead of a normal struct to avoid the pitfalls associated with having mutable structs and forgetting to pass them with the ref keyword - we can manage this with `FastBufferReader` and `FastBufferWriter` because those are internal and we can rely on our ability to understand the implications of using them as structs, but our end users may be less experienced and may inadvertently pass this forward to calls that do not take the value by ref. A `ref struct` is always passed by reference rather than copied, but does have the disadvantage of being limited to the stack frame - so we can't use it for `FastBufferWriter` and `FastBufferReader` because we persist those.
```

Comment:
I'm sorry to bother you with this (because I could do a local repro to understand this better).

Are we saying we'd accept `ref struct` only to preserve ability to mutate the source struct being passed into our methods? Is that a requirement for our readers because we do not want to alloc for types but rather write "into" the instance user passed into our reader? What if we were to split `.Serialize()` into two functions here in this case and call them `Serialize(with-no-ref-struct-param)` & `Deserialize(with-ref-struct-param)` to cover both cases?

Well, if I'm misunderstanding this thing entirely, I think a few small code snippets in this section educating me about the issue would be super helpful :)

Reply:
The reason to make it a ref struct is that the other ones (BufferReader and BufferWriter) being regular structs makes them landmines for inexperienced programmers since they have to be passed by ref always or things break (i.e., passing without the ref keyword creates a copy, which means you end up with two writers or readers pointing at the same buffer but tracking position separately and your writes get lost or overwritten). To me that's an acceptable thing for us to take on internally but not great for us to pass to our users. The suggestion to make BufferSerializer a ref struct is to disarm that landmine - ref strict S still don't have to be allocated (they're stack memory/value types), but they're passed by ref implicitly like classes are, so users can't screw up by forgetting the ref keyword. 

The downside, and reason we can't use them for Reader and Writer, is that ref structs can't outlive the stack context where they were created - they can't be boxed, can't be fields in a class or struct, can't be returned from the function, etc. When the function returns they are destroyed and that's that. But given BufferSerializer is just a wrapper class with helper methods, it doesn't need any persistence, so those limitations don't hurt it and they keep users from being able to make mistakes with it.

Reply:
I see, I think we can live with ref struct BufferSerializer, it makes sense. is there a C# doc that I could read more about this stuff? 👀

Reply:
huh, I got this: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#ref-struct